### ISSUE TITLE
3.x Store last event name

### ISSUE THREAD

**Issue by sdustinh** [cakephp/cakephp#7404]
This is a minor enhancement request but would it be possible to store the name of the last event dispatched on the event manager?

The reason for this is I was implementing something on a component that would possibly redirect a request from the startup event. This isn't a component I wanted to add on the controller's initialize method so I added it directly to the action. Something like this would be helpful:

```php
$controller = $event->subject();
$lastEvent = $controller->eventManager()->lastEvent;

if ($lastEvent === 'Controller.initialize') {
    // ...
}
```

...or even something like this...

```php
public function initialize(array $config)
{
    $controller = $this->_registry->getController();
    $eventManager = $controller->eventManager();
    
    if (in_array('Controller.startup', $eventManager->dispatchedEvents)) {
        $eventManager->onNextEvent([$this, 'methodName']);
    }
}
```

Just something I thought might be helpful as I was working.

**Comment by markstory** [comment]
Keeping a reference to the most recent event seems more reusable, as it could be used in testing. The onNextEvent() leaves me thinking

* what happens if there are no more events?
* how does the attached listener get removed?

In addition next could leade to surprising results if code is refactored or reorganized.

**Comment by sdustinh** [comment]
I could definitely be happy with a simple `lastEvent` or a stack of `dispatchedEvents`.

I hadn't seriously considered the `onNextEvent()` but the more I think about it the more I kind of want to play around with it. I wonder if something could be done with `__destruct()` or maybe a cleanup in the Dispatcher.

It might be that missing use case @lorenzo was looking for in #7358. If there was a global shutdown event that could be used to clean up leftover events and would open the door for `on<After|Before>('Controller.initialize', $callback)` or `off<After|Before>Next('Controller.shutdown', $callback)` too.

Alright. I think I'm getting carried away here...

**Comment by jippi** [comment]
in Crud we store the "events" a $event goes through - very useful for debugging to have a full stack of them

**Comment by lorenzo** [comment]
This was implemented by #8853
